fix(aria-roles): correct abstract roles (types) for aria-roles

gaiety-deque opened the pull request:
Too many aria roles were set to widget or otherwise were incorrect, added a comment source for where I got my information for which abstract roles were the type of other aria roles

Fix: #4371

gaiety-deque commented:
@WilcoFiers unsure how to test this PR's changes against the comment you left in the referenced issue, I don't see a cookie notice on the horse insurance website. Any guidance is appreciated.

@straker I feel reasonably good about this change but I'm a bit unsure if it fully addresses the original issue as it seemed fairly open ended

straker commented on lib/standards/aria-roles.js:
```diff
@@ -17,21 +18,21 @@
   },
   alertdialog: {
-    type: 'widget',
+    type: 'window',
```
`window` isn't a type we support. [Valid types](https://github.com/dequelabs/axe-core/blob/develop/doc/standards-object.md#structure-1) are:

* abstract
* widget
* structure
* landmark

So probably this should be `structure` since it's not a landmark. Also according to https://www.w3.org/WAI/ARIA/1.2/class-diagram/rdf_model.svg `alert` is a section, which is under structure. 

```suggestion
    type: 'structure',
```

gaiety-deque replied:
Despite it being an abstract role under https://www.w3.org/TR/wai-aria/#abstract_roles referencing `alertdialog` and `dialog` under https://www.w3.org/TR/wai-aria/#window_roles ?

straker replied:
window is fine

straker commented on lib/standards/aria-roles.js:
```diff
@@ -119,7 +120,7 @@ const ariaRoles = {
   },
   combobox: {
-    type: 'widget',
+    type: 'composite',
```
I think this should remain a widget. According to https://www.w3.org/WAI/ARIA/1.2/class-diagram/rdf_model.svg `combobox` is a composite, but so is a `spinbutton` and `spinbutton` is definitely a widget. So I think we're safe to say any interactable composite should be a widget.

```suggestion
    type: 'widget',
```

gaiety-deque replied:
Perhaps a good use of our 1:1 today is for us to talk about how you arrived at that conclusion, when there's two arrows coming off an arrow in the rdf model is makes it quite unclear.

straker replied:
Add comment above `type` with `// Note: spec difference` as we want to keep `widget` role

straker commented on lib/standards/aria-roles.js:
```diff
@@ -17,21 +18,21 @@
   },
   application: {
     // Note: spec difference
-    type: 'landmark',
+    type: 'structure',
```
This was an intentional difference

gaiety-deque replied:
Reverted this change 2d185f65

straker commented on lib/standards/aria-roles.js:
```diff
@@ -169,7 +170,7 @@ const ariaRoles = {
   },
   dialog: {
-    type: 'widget',
+    type: 'window',
```
Probably `structure` again as not a landmark

```suggestion
    type: 'structure',
```

gaiety-deque replied:
Since we said `window` is fine for `alertdialog` I assume it's fine here as well

straker commented on lib/standards/aria-roles.js:
```diff
@@ -301,7 +302,7 @@ const ariaRoles = {
   },
   listbox: {
-    type: 'widget',
+    type: 'composite',
```
```suggestion
    type: 'widget',
```

straker replied:
Add comment above `type` with `// Note: spec difference` as we want to keep `widget` role

straker commented on lib/standards/aria-roles.js:
```diff
@@ -1,4 +1,5 @@
 // Source: https://www.w3.org/TR/wai-aria-1.1/#roles
+// Source for abstract roles: https://www.w3.org/TR/wai-aria/#abstract_roles and https://www.w3.org/WAI/ARIA/1.2/class-diagram/rdf_model.svg
```
According to https://www.w3.org/WAI/ARIA/1.2/class-diagram/rdf_model.svg `tabpanel` is a section, which is under structure.

straker replied:
Don't reference the svg instead we make spec difference for tabpanel to not be `widget`

gaiety-deque replied:
Removed svg as comment reference 2d185f65

straker reviewed (changes requested):
Thanks for taking this on. 

https://www.w3.org/WAI/ARIA/1.2/class-diagram/rdf_model.svg is an interesting resource. I've not seen it before. It's also hard to follow sometimes what line belongs to what, so I hope I parsed it correctly.

straker commented on lib/rules/widget-not-inline-matches.js:
```diff
@@ -18,7 +18,8 @@ const matchesFns = [
 function isWidgetType(vNode) {
-  return getRoleType(vNode) === 'widget';
+  const roleType = getRoleType(vNode);
+  return roleType === 'widget' || roleType === 'composite';
```
Undo this change since we'll leave `listbox` and `combobox` as widgets

gaiety-deque replied:
Undid this change 2d185f65

straker commented:
Will also need to add exception to `focus-order-semantics` to allow `alert` and `alertdialog` roles (add unit & integration tests to match)

gaiety-deque commented:
> Will also need to add exception to `focus-order-semantics` to allow `alert` and `alertdialog` roles (add unit & integration tests to match)

@straker added an exception to `focus-order-semantics`, renamed check and added a `window` test describe for both allowed role types. Refactored the test check as well, I'm all about verbose test files with repetition but 500+ lines had me scrolling all over the place losing track so I simplified down the checks to a static easier to comprehend list facd65f6 (and [bfa95ae](https://github.com/dequelabs/axe-core/pull/4421/commits/bfa95ae2ff233de32390777f735cd35ef1e50fa4))

WilcoFiers commented on lib/standards/aria-roles.js:
```diff
@@ -17,13 +18,13 @@
 const ariaRoles = {
   alert: {
-    type: 'widget',
+    type: 'structure',
```
This creates a problem for `focus-order-semantics`. With this change, the following is going to start failing. I don't think we should want that.

```html
<div role="alert" tabindex="0"> Hello world </div>
```

This is what we talked about on Thursday. We'll need some way to exempt these roles that are no longer considered as widgets in the `focus-order-semantics` rule. I think the best way to do that is just to add them to the `VALID_ROLES_FOR_SCROLLABLE_REGIONS` object in `valid-scrollable-semantics-evaluate`.

WilcoFiers commented on lib/standards/aria-roles.js:
```diff
@@ -816,7 +819,7 @@ const ariaRoles = {
   },
   timer: {
-    type: 'widget',
+    type: 'structure',
```
When you're adding roles to `valid-scrollable-semantics`, I think you can leave this one out. `role=timer` doesn't seem like it should be focusable.

WilcoFiers commented on lib/standards/aria-roles.js:
```diff
@@ -340,7 +343,7 @@ const ariaRoles = {
   },
   marquee: {
-    type: 'widget',
+    type: 'structure',
```
When you're adding roles to `valid-scrollable-semantics`, I think you can leave this one out. `role=marquee` doesn't seem like it should be focusable.

WilcoFiers commented on test/checks/aria/has-widget-or-window-role.js:
```diff
@@ -0,0 +1,118 @@
+    axe.testUtils
+      .getCheckEvaluate('has-widget-or-window-role')
+      .call(checkContext, currentNode);
+  const roles = {
```
This feels a little brittle. Generally what we do is we have the unit tests just check the code works. One or two examples of each role type is enough. And then we have fairly extensive integration tests. Those are a little more telling.

gaiety-deque replied:
I'll be reverting this change since the desire is to not change this check.

Though I believe this change was not brittle. This was a refactor from ~500 lines of tests to ~100 that checked the same things, but it was difficult to parse by a person before. I found myself getting quite lost in the code previously. The underlying test code was the same with this change.

WilcoFiers reviewed (changes requested):
Can you add integration tests for:
1. Show that target-size no longer applies to some of the things that are no longer widgets. Most importantly alertdialog
2. Show that target-size no longer treats elements with those roles as neighbor targets
3. Show which roles are still allowed to be focusable under valid-scrollable-semantics, and which are not

gaiety-deque commented on lib/checks/aria/has-widget-or-window-role-evaluate.js:
```diff
@@ -0,0 +1,26 @@
+ * @memberof checks
+ * @return {Boolean} True if the element uses a `widget`, `composite`, or `window` abstract role (type). False otherwise.
+ */
+// # TODO: change to abstract role for widget and window
```
Not intentional, reverted a07c488
